File: python/bench.py
from random import randint
import string
import os
from twisted.internet import reactor
from twisted.internet.defer import inlineCallbacks
from twisted.enterprise import adbapi
from cyclone.web import Application, RequestHandler

reactor.suggestThreadPoolSize(128)

letters = list(string.ascii_letters + ' ' + string.digits)
all_chars = string.maketrans('', '')
result_chars = ''.join(letters[b % len(letters)] for b in range(len(all_chars)))
trans = string.maketrans(all_chars, result_chars)

def randstr(size):
    return os.urandom(size).translate(trans)

class BenchEndpoint(RequestHandler):
    def initialize(self, db=None):
        self.db = db

    @inlineCallbacks
    def get(self):
        coin = randint(0, 9)
        if coin < 6:
            result = yield self.db.runQuery("SELECT title, thumb, nc, nv FROM tst ORDER BY id DESC LIMIT 10")
            data = result and [dict(title=r[0], thumb=r[1], nc=r[2], nv=r[3]) for r in result] or None
            resp = dict(list=data)

        elif coin < 8:
            title = randstr(140)
            thumb = randstr(140)
            nc = randint(0, 1000)
            nv = randint(0, 5000)
            yield self.db.runOperation("INSERT INTO tst(title, thumb, nc, nv) VALUES (%s, %s, %s, %s)", 
                                       (title, thumb, nc, nv,))
            resp = dict(action="insert", title=title, thumb=thumb, nc=nc, nv=nv)

        else:
            title = randstr(140)
            thumb = randstr(140)
            nc = randint(0, 1000)
            nv = randint(0, 5000)
            yield self.db.runOperation("UPDATE tst SET title=%s, thumb=%s, nc=%s, nv=%s "
                                       "WHERE id=(SELECT max(id) as m FROM tst LIMIT 1)",
                                       (title, thumb, nc, nv,))
            resp = dict(action="update", title=title, thumb=thumb, nc=nc, nv=nv)
        
        self.write(resp)


class Bench(Application):
    def __init__(self):
        self.db_pool = adbapi.ConnectionPool("psycopg2", host="127.0.0.1", database="bench",
                                             user="bench", password="bench", cp_max=32, cp_reconnect=True)

        handlers = [
            (r"/bench", BenchEndpoint, dict(db=self.db_pool)),
        ]
        Application.__init__(self, handlers, dict(xheaders=True))

    def log_request(self, handler):
        pass

    def stopFactory(self):
        self.db_pool.close()

File: node/bench.js
var http = require('http');
var pg = require('pg').native;

var config = {
  user: 'bench',
  database: 'bench',
  password: 'bench',
  max: 32,
};

var pool = new pg.Pool(config);

function randInt(max) {
  return Math.floor(Math.random() * (max + 1));
}

function randStr(length) {
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    for(var i=0; i < length; i++)
        text += possible.charAt(Math.floor(Math.random() * possible.length));

    return text;
}

function query(callback) {
  pool.connect(function(err, client, done) {
    client.query('SELECT * FROM tst ORDER BY id DESC LIMIT 10', [], function(err, result) {
      done();
      callback({list: result.rows});
    });
  });
}

function insert(callback) {
  pool.connect(function(err, client, done) {
    var title = randStr(140);
    var thumb = randStr(140);
    var nc = randInt(1000);
    var nv = randInt(5000);
    client.query('INSERT INTO tst(title, thumb, nc, nv) VALUES($1, $2, $3, $4)', 
        [title, thumb, nc, nv], 
        function(err, result) {
            done();
            callback({action: "insert", title: title, thumb: thumb, nc: nc, nv: nv});
        });
   });
}

function update(callback) {
  pool.connect(function(err, client, done) {
    var title = randStr(140);
    var thumb = randStr(140);
    var nc = randInt(1000);
    var nv = randInt(5000);
    client.query('UPDATE tst SET title=$1, thumb=$2, nc=$3, nv=$4 WHERE id=(SELECT max(id) FROM tst)', 
        [title, thumb, nc, nv], 
        function(err, result) {
            done();
            callback({action: "update", title: title, thumb: thumb, nc: nc, nv: nv});
        });
   });
}

function response(resp, data) {
  resp.writeHead(200, {'Content-Type': 'application/json'});
  resp.end(JSON.stringify(data));
}

function handler(req, res) {
  var coin = randInt(9);
  if (coin < 6) {
    query(function(data) { response(res, data); });
  } else if (coin < 8) {
    insert(function(data) { response(res, data); });
  } else {
    update(function(data) { response(res, data); });
  }
}

var server = http.createServer(handler);
server.listen(8080);
